File: 22.07.26(16946).py
import math
import sys
from collections import deque
input = sys.stdin.readline
n, m = map(int,input().split())
visit = [[False] * m for i in range(n)]
dir = [(0,-1), (1, 0), (0, 1), (-1, 0)]
zeros = [[0] * m for i in range(n)]
ccnntt = {}
idx  =  1
board = [list(map(int, input().strip())) for i in range(n)]
def BFS(y, x):
    q = deque()
    visit[y][x] = True
    q.append([y, x])
    cnt = 1
    while q:
        h_y, h_x = q.pop()
        zeros[h_y][h_x] = idx
        for i, j in dir:
            n_y, n_x = h_y + i, h_x + j
            if n_y < 0 or n_x < 0 or n_x >= m or n_y >= n or visit[n_y][n_x] == True or board[n_y][n_x] != 0:continue
            cnt+=1
            visit[n_y][n_x] = True
            q.append([n_y, n_x])
    return cnt
def find(y, x):
    s = set()
    cnt = 1
    for i, j in dir:
        n_y, n_x = y + i, x + j
        if n_y < 0 or n_x < 0 or n_x >= m or n_y >= n or not zeros[n_y][n_x]:continue
        s.add(zeros[n_y][n_x])
    #print(s)
    for i in s:
        cnt += ccnntt[i]
    return cnt%10



for i in range(n):
    for j in range(m):
        if visit[i][j] == False and board[i][j] == 0:
            k = BFS(i, j)
            ccnntt[idx] = k
            idx+=1
answer = [[0] * m for i in range(n)]
for i in range(n):
    for j in range(m):
        if board[i][j]:
            answer[i][j] = find(i,j)
for i in range(n):
    for j in answer[i]:
        print(j ,end='')
    print()

File: 22.07.26(16946).java
import java.awt.desktop.PreferencesEvent;
import java.awt.image.AreaAveragingScaleFilter;
import java.io.*;
import java.nio.CharBuffer;
import java.time.chrono.ThaiBuddhistEra;
import java.util.*;

class P{
    public int y,x;
    P(int y,int x){
        this.y = y;
        this.x = x;
    }
}


public class Main {
    static int n, m, idx;
    static ArrayList<Integer> groupCount;
    static int []dy = {0,0,-1,1};
    static int []dx = {-1,1,0,0};
    static int [][]group;
    static int [][]board;
    static StringBuilder sb = new StringBuilder();
    static boolean [][] visit;
    static int DFS(int y, int x, int idx){
        Queue <P> q = new LinkedList<>();
        q.add(new P(y,x));
        visit[y][x] = true;
        group[y][x] = idx;
        int cnt =1;
        while(!q.isEmpty()) {
            int h_x = q.peek().x;
            int h_y = q.peek().y;
            q.remove();
            for (int i = 0; i < 4; i++) {
                int n_x = h_x + dx[i];
                int n_y = h_y + dy[i];
                if (n_x < 0 || n_x >= m || n_y < 0 || n_y >= n || visit[n_y][n_x] == true || board[n_y][n_x] != 0)continue;
                visit[n_y][n_x] = true;
                q.add(new P(n_y, n_x));
                group[n_y][n_x] = idx;
                cnt++;
            }}
            return cnt;
        }

    public static void main(String args[]) throws IOException {
        Scanner sc = new Scanner(System.in);
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        visit = new boolean[n][m];
        board = new int[n][m];
        groupCount = new ArrayList<>();
        group = new int[n][m];
        groupCount.add(0);
        idx = 1;
        //n=sc.nextInt();
        for(int i=0; i<n; i++){
            Arrays.fill(visit[i], false);
        }
        for(int i=0; i<n; i++){
            String[] s  = br.readLine().split("");
            for(int j=0; j<m; j++){
                board[i][j] = Integer.parseInt(s[j]);
            }
        }
        for(int i =0 ;i<n; i++){
            for(int j=0; j<m; j++){
                if(visit[i][j] == false && board[i][j] == 0){
                    groupCount.add(DFS(i,j,idx++));
                }
            }
        }

        for(int i=0; i<n; i++){
            for(int j = 0; j<m; j++){
                find(i,j);
            }
            sb.append("\n");
        }
        System.out.println(sb);
        //bw.write(sb.toString());
    }
    static void find(int y, int x){
        if(board[y][x] == 0){
            sb.append("0");
            return;
        }
        HashSet<Integer> set  = new HashSet<>();
        for(int i=0; i<4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(nx <0 || nx >=m || ny<0 || ny >=n)continue;
            set.add(group[ny][nx]);
        }
        int cnt = 1;
        for(int e : set){
            cnt += groupCount.get(e);
        }
        cnt %= 10;
        sb.append(cnt);
    }
}
